### ISSUE TITLE
Bug: Setting seed in every reset 

### ISSUE THREAD

**Issue by sai-prasanna** [weipu-zhang/STORM#8]
I found this bug when I used the STORM code for my experiments. The environment seed is being set in every reset. This is incorrect; it will lead to the same start state being sampled for every episode. I am not sure how much this impacts results in ATARI, but if any of the environments where you report the scores have a start state distribution, this has to be changed, and it has to be re-run.

https://github.com/weipu-zhang/STORM/blob/e0b3fd44320d7e213ec905c673ad3f35b61b89f4/env_wrapper.py#L40

This could have an even bigger impact if some Atari games have randomness in the transition dynamics. Essentially, all the trajectories will be deterministic!!

**Comment by zzzc18** [comment]
Hi,

This issue is mainly due to historical reasons. By the time we built this repo we couldn't find a standard wrapper for Atari 100k. And the gymnasium documentation about seed setting was vague. (Like should one set the seed only at the beginning or at each episode?) So we decided to choose one of the possible settings. 

For the randomness, it seems that only the scene distribution in PrivateEye is affected by the seed a lot. Other random behaviours are still random even if the seed is fixed.

Now we suggest using a wrapper like the following and using v4 environments (like IRIS/DIAMOND/EfficientZero). We've tried this setting for another project. The results for different environments shift, but the overall performance is similar. 


```python
class StochasticSeedEnvWrapper(gymnasium.Wrapper):
    def __init__(self, env):
        super().__init__(env)

    def reset(self, **kwargs):
        kwargs["seed"] = np.random.randint(2**32)
        obs, _ = self.env.reset(**kwargs)
        return obs, _

    def step(self, action):
        return self.env.step(action)

def build_atari_env(env_name, seed, image_size=None):
    env = gymnasium.make(env_name, full_action_space=False, render_mode="rgb_array", frameskip=1)
    env = StochasticSeedEnvWrapper(env)
    env = env_wrapper.NoopResetWrapper(env, noop_max=30)
    env = env_wrapper.MaxLast2FrameSkipWrapper(env, skip=4)
    if image_size is not None:
        env = gymnasium.wrappers.ResizeObservation(env, shape=image_size)
    env = env_wrapper.LifeLossInfoWrapper(env)
    env = env_wrapper.SqueezeActionDimWrapper(env)
    env = gymnasium.wrappers.TimeLimit(env, max_episode_steps=5000)

    return env

build_atari_env("PongNoFrameskip-v4", 1, (64, 64))
```

**Comment by weipu-zhang** [comment]
Hi,

This issue is mainly due to historical reasons. By the time we built this repo we couldn't find a standard wrapper for Atari 100k. And the gymnasium documentation about seed setting was vague. (Like should one set the seed only at the beginning or at each episode?) So we decided to choose one of the possible settings. 

For the randomness, it seems that only the scene distribution in PrivateEye is affected by the seed a lot. Other random behaviours are still random even if the seed is fixed.

Now we suggest using a wrapper like the following and using v4 environments (like IRIS/DIAMOND/EfficientZero). We've tried this setting for another project. The results for different environments shift, but the overall performance is similar. 


```python
class StochasticSeedEnvWrapper(gymnasium.Wrapper):
    def __init__(self, env):
        super().__init__(env)

    def reset(self, **kwargs):
        kwargs["seed"] = np.random.randint(2**32)
        obs, _ = self.env.reset(**kwargs)
        return obs, _

    def step(self, action):
        return self.env.step(action)

def build_atari_env(env_name, seed, image_size=None):
    env = gymnasium.make(env_name, full_action_space=False, render_mode="rgb_array", frameskip=1)
    env = StochasticSeedEnvWrapper(env)
    env = env_wrapper.NoopResetWrapper(env, noop_max=30)
    env = env_wrapper.MaxLast2FrameSkipWrapper(env, skip=4)
    if image_size is not None:
        env = gymnasium.wrappers.ResizeObservation(env, shape=image_size)
    env = env_wrapper.LifeLossInfoWrapper(env)
    env = env_wrapper.SqueezeActionDimWrapper(env)
    env = gymnasium.wrappers.TimeLimit(env, max_episode_steps=5000)

    return env

build_atari_env("PongNoFrameskip-v4", 1, (64, 64))
```

**Comment by sai-prasanna** [comment]
Thanks for your comment. The behaviours will seem random, but the world model and actor-critic will easily overfit as the transition dynamics are deterministic when you set the same seed. Hence, each episode will be deterministic for the same set of actions. The training curves will overestimate the algorithm's performance in such a case.

 This happened to me in another environment where I was using your code. So, if you had reported training curves in the paper, I would strongly suggest re-running all the benchmarks.
 
 
I think the best practice is just to reset it with the seed only once during environment creation, without a wrapper. This would have stronger guarantees of reproducibility of the environment trajectories.
